split static condition tests from net_match_config()

yuwata opened the pull request:
Strictly speaking, this breaks backward compatibility of conditions in `[Match]` section; E.g.,
```
[Match]
Host=hogehoge
Host=foover
```
Previously, the setting `Host=hogehoge` was ignored, and the last entry was always used. This makes all conditions are examined, and the results are ANDed. So, no host cannot match the above condition.

To override the preexisting conditions, empty string assignment is necessary, that is,
```
[Match]
Host=hogehoge
Host=
Host=foover
```
Then the condition `Host=hogehoge` is cleared and only `Host=foover` is used.

---
The backward incompatibility is dropped from this PR.

keszybz commented on man/systemd.link.xml:
```diff
@@ -119,7 +119,9 @@
           ID of the host. See <varname>ConditionHost=</varname> in
           <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry>
-          for details.</para>
+          for details. When prefixed with an exclamation mark (<literal>!</literal>), the result is nagated.
```
"negated"

keszybz commented on man/systemd.link.xml:
```diff
@@ -130,7 +132,9 @@
           <varname>ConditionVirtualization=</varname> in
           <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry>
-          for details.</para>
+          for details. When prefixed with an exclamation mark (<literal>!</literal>), the result is nagated.
```
"negated"

keszybz commented on man/systemd.link.xml:
```diff
@@ -140,28 +144,32 @@
-            </para>
-          </listitem>
-        </varlistentry>
+          for details. When prefixed with an exclamation mark (<literal>!</literal>), the result is nagated.
```
"negated"

keszybz commented on man/systemd.link.xml:
```diff
@@ -140,28 +144,32 @@
+        <term><varname>KernelVersion=</varname></term>
+        <listitem>
+          <para>Checks whether the kernel version (as reported by <command>uname -r</command>) matches a certain
+          expression (or if prefixed with the exclamation mark does not match it). See
```
This information about negation is now repeated needlessly. I think the part in parentheses should be removed.

keszybz commented on man/systemd.link.xml:
```diff
@@ -140,28 +144,32 @@
+          expression (or if prefixed with the exclamation mark does not match it). See
+          <varname>ConditionKernelVersion=</varname> in
+          <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry> for
+          details. When prefixed with an exclamation mark (<literal>!</literal>), the result is nagated.
```
"negated"

keszybz commented on man/systemd.link.xml:
```diff
@@ -140,28 +144,32 @@
+          architecture. See <varname>ConditionArchitecture=</varname> in
           <citerefentry><refentrytitle>systemd.unit</refentrytitle><manvolnum>5</manvolnum></citerefentry>
-          for details.</para>
+          for details. When prefixed with an exclamation mark (<literal>!</literal>), the result is nagated.
```
"negated", here and below.

yuwata commented:
Thanks 😊 I will fix them soon.

yuwata commented:
> I think we can do this. But if multiple values are assigned, I think we should at least emit a warning. It's very unlikely that this is intentional.

OK. I did not think it was necessary, but this breaks a backward compatibility, so it is reasonable.

keszybz commented:
So this should go in first, then #11861, and then #11893. So I'll wait with reviews of the other ones for the update here.

yuwata commented:
Updated.

> I think we can do this. But if multiple values are assigned, I think we should at least emit a warning. It's very unlikely that this is intentional.

After re-considering this, if a unit has multiple conditions and that is intentional, then the warning is noisy and may confuse people...
I think we should warn only when the condition is always true or always false, e.g.,
```
Host=hoge
Host=foo
```
But implementing such a check is out of the scope of this PR. So, I've dropped the multiple assignment support, and I'd like to leave that for another PR. Though, I am not sure people really wants such a feature...

keszybz commented:
Needs a rebase now.

yuwata commented:
@keszybz Thank you for the review. Rebased. PTAL.
